src/git/git.go

```diff
@@ -0,0 +1,149 @@
+	destinationDirectory string
+	sshKey               string
+	replace              bool
+	repo                 *git.Repository
+	authorization        *ssh.PublicKeys
+}
+
+// GetPublicKey aims to read ssh key file and returns a go git PublicKey
+// Args:
+// 		sshKey ([]byte): Byte arry containing the ssh key.
+// 		Note: Set up this key in the git settings and pass as a parameter to the CLI tool.
+// Returns:
+// 		error: Returns an error whenever unexpected issues happen.
+func (g *GitVersioning) getPublicKey(sshKey []byte) error {
+	var publicKey *ssh.PublicKeys
+
+	publicKey, err := ssh.NewPublicKeys("git", []byte(sshKey), "")
+	if err != nil {
+		g.log.Error(err.Error())
+		return err
+	}
+
+	g.authorization = publicKey
+	return nil
+}
+
+// FormatSSHKey aims to properlly format the ssh private key.
+// It is necessary once we cannot pass a multiline variable to go build. So it is not possible
+// to do docker build --build-arg PRIVATE_KEY=$(PRIVATE_KEY) once this kind of file has several line breaks.
+// Args:
+// 		lineBreaker (string): Line break holder character. I.e.: #
+// Returns:
+// 		string: Returns the sshKey string with lineBreaker character replaced by line breaks.
+func (g *GitVersioning) formatSSHKey(lineBreaker string) string {
+	return strings.Replace(g.sshKey, lineBreaker, "\n", -1)
+}
+
+func (g *GitVersioning) validate() error {
+	if g.url == "" {
+		return errors.New("url cannot be empty")
+	}
+
+	if g.destinationDirectory == "" {
+		return errors.New("destination directory cannot be empty")
+	}
+
+	if g.sshKey == "" {
+		return errors.New("ssh key cannot be empty")
+	}
+
+	return nil
+}
+
+// cloneRepoToDirectory aims to clone the repository from remote to local.
+// Args:
+// 		auth (*ssh.PublicKeys): Authorization key.
+// Returns:
+// 		*git.Repository: Returns a repository reference.
+// 		err: Error whenever unexpected issues happen.
+func (g *GitVersioning) cloneRepoToDirectory() (*git.Repository, error) {
+	defer g.printElapsedTime("CloneRepoToDirectory")()
+
+	if err := g.validate(); err != nil {
+		g.log.Error(err.Error())
+		return nil, err
+	}
+
+	err := g.getPublicKey([]byte(g.formatSSHKey("#")))
+	if err != nil {
+		g.log.Error("error while getting public key due to: %s", err)
+		return nil, err
+	}
+
+	g.log.Info(colorYellow+"cloning repo "+colorCyan+" %s "+colorYellow+" into "+colorCyan+"%s"+colorReset, g.url, g.destinationDirectory)
+	repo, err := git.PlainClone(g.destinationDirectory, false, &git.CloneOptions{
+		Progress: os.Stdout,
+		URL:      g.url,
+		Auth:     g.authorization,
+	})
+	if err != nil {
+		if err == git.ErrRepositoryAlreadyExists {
+			g.log.Warn("repository was already cloned")
+			if g.replace {
+				g.log.Info("removing path %s", g.destinationDirectory)
+				err := os.RemoveAll(g.destinationDirectory)
+				if err != nil {
+					return nil, err
+				}
+				g.replace = false
+				return g.cloneRepoToDirectory()
+			}
+			return git.PlainOpen(g.destinationDirectory)
+		} else {
```

Comment:
do you need this else? you already have return... @esequielvirtuoso

Reply:
Not necessary.
I've removed it.